File: results/1.0/GZIP/gzip-deepseek-ai-deepseek-v3.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

HParser *gzip_parser() {
    HParser *id1 = h_uint8();
    HParser *id2 = h_uint8();
    HParser *cm = h_uint8();
    HParser *flg = h_uint8();
    HParser *mtime = h_uint32();
    HParser *xfl = h_uint8();
    HParser *os = h_uint8();

    HParser *extra = h_optional(h_sequence(h_uint16(), h_length_value(h_uint16(), h_many(h_uint8())), NULL));
    HParser *fname = h_optional(h_sequence(h_uint8(), h_many(h_uint8()), NULL));
    HParser *fcomment = h_optional(h_sequence(h_uint8(), h_many(h_uint8()), NULL));
    HParser *crc16 = h_optional(h_uint16());

    HParser *compressed_data = h_many(h_uint8());
    HParser *crc32 = h_uint32();
    HParser *isize = h_uint32();

    HParser *gzip_header = h_sequence(id1, id2, cm, flg, mtime, xfl, os, extra, fname, fcomment, crc16, NULL);
    HParser *gzip_file = h_sequence(gzip_header, compressed_data, crc32, isize, NULL);

    return gzip_file;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <binary_file>\n", argv[0]);
        return 1;
    }

    FILE *file = fopen(argv[1], "rb");
    if (!file) {
        perror("Failed to open file");
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);

    uint8_t *data = (uint8_t *)malloc(file_size);
    if (!data) {
        perror("Failed to allocate memory");
        fclose(file);
        return 1;
    }

    fread(data, 1, file_size, file);
    fclose(file);

    HParser *gzip = gzip_parser();
    HParseResult *result = h_parse(gzip, data, file_size);

    if (result) {
        printf("GZIP file parsed successfully\n");
        h_parse_result_free(result);
    } else {
        printf("Failed to parse GZIP file\n");
    }

    free(data);
    return 0;
}

File: results/1.0/GZIP/gzip-deepseek-ai-deepseek-v3.rs
use std::fs::File;
use std::io::Read;
use nom::{
    bytes::complete::{tag, take},
    combinator::{map, map_res},
    multi::length_data,
    number::complete::{le_u16, le_u32, u8},
    sequence::tuple,
    IResult,
};
use std::str;

#[derive(Debug)]
struct GzipHeader {
    id1: u8,
    id2: u8,
    cm: u8,
    flg: u8,
    mtime: u32,
    xfl: u8,
    os: u8,
    xlen: Option<u16>,
    extra: Option<Vec<u8>>,
    fname: Option<String>,
    fcomment: Option<String>,
    hcrc: Option<u16>,
}

#[derive(Debug)]
struct GzipFooter {
    crc32: u32,
    isize: u32,
}

fn parse_gzip_header(input: &[u8]) -> IResult<&[u8], GzipHeader> {
    let (input, (id1, id2, cm, flg, mtime, xfl, os)) = tuple((u8, u8, u8, u8, le_u32, u8, u8))(input)?;
    let (input, xlen) = if flg & 0x04 != 0 {
        map(le_u16, Some)(input)?
    } else {
        (input, None)
    };
    let (input, extra) = if let Some(len) = xlen {
        map(take(len as usize), |v: &[u8]| Some(v.to_vec()))(input)?
    } else {
        (input, None)
    };
    let (input, fname) = if flg & 0x08 != 0 {
        map(map_res(length_data(u8), |v: &[u8]| str::from_utf8(v).map(|s| s.to_string())), Some)(input)?
    } else {
        (input, None)
    };
    let (input, fcomment) = if flg & 0x10 != 0 {
        map(map_res(length_data(u8), |v: &[u8]| str::from_utf8(v).map(|s| s.to_string())), Some)(input)?
    } else {
        (input, None)
    };
    let (input, hcrc) = if flg & 0x02 != 0 {
        map(le_u16, Some)(input)?
    } else {
        (input, None)
    };
    Ok((
        input,
        GzipHeader {
            id1,
            id2,
            cm,
            flg,
            mtime,
            xfl,
            os,
            xlen,
            extra,
            fname,
            fcomment,
            hcrc,
        },
    ))
}

fn parse_gzip_footer(input: &[u8]) -> IResult<&[u8], GzipFooter> {
    let (input, (crc32, isize)) = tuple((le_u32, le_u32))(input)?;
    Ok((input, GzipFooter { crc32, isize }))
}

fn parse_gzip(input: &[u8]) -> IResult<&[u8], (GzipHeader, Vec<u8>, GzipFooter)> {
    let (input, header) = parse_gzip_header(input)?;
    let compressed_data_end = input.len() - 8;
    let (input, compressed_data) = take(compressed_data_end)(input)?;
    let (input, footer) = parse_gzip_footer(input)?;
    Ok((input, (header, compressed_data.to_vec(), footer)))
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <file>", args[0]);
        return;
    }

    let mut file = File::open(&args[1]).expect("Failed to open file");
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer).expect("Failed to read file");

    match parse_gzip(&buffer) {
        Ok((_, (header, compressed_data, footer))) => {
            println!("Header: {:?}", header);
            println!("Compressed Data Length: {}", compressed_data.len());
            println!("Footer: {:?}", footer);
        }
        Err(e) => eprintln!("Failed to parse GZIP file: {:?}", e),
    }
}